src/main/java/com/pinterest/secor/consumer/Consumer.java

```diff
@@ -78,6 +82,24 @@ private void init() throws Exception {
         mMessageParser = ReflectionUtil.createMessageParser(mConfig.getMessageParserClass(), mConfig);
         mMessageTransformer = ReflectionUtil.createMessageTransformer(mConfig.getMessageTransformerClass(), mConfig);
         mUnparsableMessages = 0.;
+
+        mUploadOnShutdown = mConfig.getUploadOnShutdown();
+        if (mUploadOnShutdown) {
+            Runtime.getRuntime().addShutdownHook(this.new ShutdownHook());
+        }
+    }
+
+    // When the JVM starts to shut down, tell the Consumer thread to upload once and wait for it to finish.
+    private class ShutdownHook extends Thread {
```

Comment:
Maybe the class name should be called FinalUploadShutdownHook to make the purpose more explicit

Reply:
Done